src/control/system/database_test.go

```diff
@@ -717,3 +717,93 @@ func TestSystem_Database_OnEvent(t *testing.T) {
 		})
 	}
 }
+
+func TestSystem_Database_GroupMap(t *testing.T) {
+	membersWithStates := func(states ...MemberState) []*Member {
+		members := make([]*Member, len(states))
+
+		for i, ms := range states {
+			members[i] = MockMember(t, uint32(i), ms)
+		}
+
+		return members
+	}
+
+	for name, tc := range map[string]struct {
+		members     []*Member
+		expGroupMap *GroupMap
+		expErr      error
+	}{
+		"empty membership": {
+			expErr: ErrEmptyGroupMap,
+		},
+		"excluded members not included": {
+			// This is a bit fragile, but I don't see a better way to maintain
+			// this list. We'll just need to keep it updated as the states change.
+			members: membersWithStates(
+				MemberStateUnknown, MemberStateAwaitFormat,
+				MemberStateStarting, MemberStateReady, MemberStateJoined,
+				MemberStateStopping, MemberStateStopped,
+				MemberStateEvicted, MemberStateExcluded,
+				MemberStateErrored, MemberStateUnresponsive,
+			),
+			expGroupMap: &GroupMap{
+				Version: 11,
+				RankURIs: map[Rank]string{
+					0:  mockControlAddr(t, 0).String(),
+					2:  mockControlAddr(t, 2).String(),
+					3:  mockControlAddr(t, 3).String(),
+					4:  mockControlAddr(t, 4).String(),
+					5:  mockControlAddr(t, 5).String(),
+					6:  mockControlAddr(t, 6).String(),
+					9:  mockControlAddr(t, 9).String(),
```

Comment:
Took me a moment to correlate the numbers with the states. Okay as-is, but I wonder if there's a way to make the relationship a little easier to understand.

Reply:
Updated.